batchflow/models/torch/base_mixins.py

```diff
@@ -221,6 +222,197 @@ class OptimalBatchSizeMixin:
+                            start_batch_size=4, max_iters=25, pbar='n',
+                            method='train', max_memory=100, frequency=0.01,
+                            tail_size=20):
+        count = 0
+
+        # if None => make equal distance between low, start_batch_size and high
+        high = high if high is not None else 2 * start_batch_size - low
+
+        low = low if isinstance(low, list) else [low]
+        high = high if isinstance(high, list) else [high]
+        batch_size = start_batch_size if isinstance(start_batch_size, list) else [start_batch_size]
+
+        n_iters = int(np.ceil(np.log2(high[-1] - low[-1])))
+        notifier = Notifier(n_iters=n_iters, bar=pbar,
+                            monitors=[{'source': low, 'name': 'low'},
+                                      {'source': high, 'name': 'high'},
+                                      {'source': batch_size, 'name': 'batch_size'}])
+        while True:
+            try:
+                notifier.update()
+                # exit when >= max_iters
+                if count >= max_iters:
+                    batch_size.append(low[-1])
+                    break
+                count += 1
+
+                # monitor consumed memory
+                with GPUMemoryMonitor(frequency=frequency) as monitor:
+                    input = inputs or self.make_placeholder_data(batch_size[-1], to_device=False)
+                    input = list(input) if isinstance(input, (tuple, list)) else [input]
+                    input = [item[:batch_size[-1]] for item in input]
+
+                    if method == 'train':
+                        target = targets or self.predict(inputs=input, outputs='predictions')
+                        target = list(target) if isinstance(target, (tuple, list)) else [target]
+                        target = [item[:batch_size[-1]] for item in target]
+
+                        _ = self.train(inputs=input, targets=target, microbatch_size=False)
+                    else:
+                        _ = self.predict(inputs=input, microbatch_size=False)
+
+                data = monitor.data
+                # take mean of top_k memory measures
+                consumed_memory = np.mean(np.sort(data, axis=0)[-tail_size:])
+
+                # update borders
+                if consumed_memory > max_memory:
+                    high.append(batch_size[-1])
+                else:
+                    low.append(batch_size[-1])
+
+                batch_size.append((high[-1] + low[-1]) // 2)
+            except RuntimeError as exception:
+                if self.is_oom_error(exception):
+                    high.append(batch_size[-1])
+                    batch_size.append((high[-1] + low[-1]) // 2)
+                else:
+                    raise # some other error not memory related
+
+            if high[-1] - low[-1] <= 1:
+                break
+
+            self.garbage_collection_cuda()
```

Comment:
move this before the `if/break`: this way, you would not need to call `gc` after the for-loop